Build failing after 16.1.2 to 16.2 upgrade

### Link to the code that reproduces this issue

https://github.com/yukosgiti/next-js-16-2-0-bug

### To Reproduce

The reproduction has 2 commits. 
1st commit - pdfjs working normally in a server action.
2nd commit - ran npx next upgrade

- checkout the 1st commi
 - npm install
- npm run dev
- click on button at /
- pdf created.
- then run: npx next upgrade
- the dev server errors

### Current vs. Expected behavior

the website should work as it did before the upgrade

The build and the dev server gives the following error:

```ts
GET / 500 in 32ms (next.js: 19ms, application-code: 13ms)
[browser] Uncaught Error: ./node_modules/fflate/lib/node.cjs:22:13
Module not found: Can't resolve <dynamic>
  20 | node_worker_1["default"] = Worker ? function (c, _, msg, transfer, cb) {
  21 |     var done = false;
> 22 |     var w = new Worker(c + workerAdd, { eval: true })
     |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  23 |         .on('error', function (e) { return cb(e, null); })
  24 |         .on('message', function (m) { return cb(null, m); })
  25 |         .on('exit', function (c) {



Import trace:
  Server Component:
    ./node_modules/fflate/lib/node.cjs
    ./node_modules/jspdf/dist/jspdf.node.min.js
    ./app/create-pdf-action.ts
    ./app/page.tsx

https://nextjs.org/docs/messages/module-not-found


    at <unknown> (Error: ./node_modules/fflate/lib/node.cjs:22:13)
    at <unknown> (https://nextjs.org/docs/messages/module-not-found)
    at <unknown> (https://nextjs.org/docs/messages/module-not-found)
[browser] ./node_modules/fflate/lib/node.cjs:22:13
Module not found: Can't resolve <dynamic>
  20 | node_worker_1["default"] = Worker ? function (c, _, msg, transfer, cb) {
  21 |     var done = false;
> 22 |     var w = new Worker(c + workerAdd, { eval: true })
     |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  23 |         .on('error', function (e) { return cb(e, null); })
  24 |         .on('message', function (m) { return cb(null, m); })
  25 |         .on('exit', function (c) {

Import trace:
  Server Component:
    ./node_modules/fflate/lib/node.cjs
    ./node_modules/jspdf/dist/jspdf.node.min.js
    ./app/create-pdf-action.ts
    ./app/page.tsx

https://nextjs.org/docs/messages/module-not-found 
    at console.overrideMethod [as error] (chrome-extension://fmkadmapgofadopljbjfkapdkoienihi/build/installHook.js:1:174822) (chrome-extension://fmkadmapgofadopljbjfkapdkoienihi/build/installHook.js:1:174822)
[browser] ./node_modules/fflate/lib/node.cjs:22:13
Module not found: Can't resolve <dynamic>
  20 | node_worker_1["default"] = Worker ? function (c, _, msg, transfer, cb) {
  21 |     var done = false;
> 22 |     var w = new Worker(c + workerAdd, { eval: true })
     |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  23 |         .on('error', function (e) { return cb(e, null); })
  24 |         .on('message', function (m) { return cb(null, m); })
  25 |         .on('exit', function (c) {

Import trace:
  Server Component:
    ./node_modules/fflate/lib/node.cjs
    ./node_modules/jspdf/dist/jspdf.node.min.js
    ./app/create-pdf-action.ts
    ./app/page.tsx

https://nextjs.org/docs/messages/module-not-found 
    at console.overrideMethod [as error] (chrome-extension://fmkadmapgofadopljbjfkapdkoienihi/build/installHook.js:1:174822) (chrome-extension://fmkadmapgofadopljbjfkapdkoienihi/build/installHook.js:1:174822)
```

### Provide environment information

```bash
Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 25.3.0: Wed Jan 28 20:56:42 PST 2026; root:xnu-12377.91.3~2/RELEASE_ARM64_T8142
  Available memory (MB): 16384
  Available CPU cores: 10
Binaries:
  Node: 22.21.1
  npm: 11.6.4
  Yarn: 1.22.22
  pnpm: 10.26.0
Relevant Packages:
  next: 16.2.0 // Latest available version is detected (16.2.0).
  eslint-config-next: N/A
  react: 19.2.4
  react-dom: 19.2.4
  typescript: 5.9.3
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

React, Server Actions

### Which stage(s) are affected? (Select all that apply)

next dev (local), next build (local)

### Additional context

It might be an issue with react as well.

I'm trying to reproduce this on my side

On Thu, Mar 19, 2026, 9:33 AM Yüksel Kapan ***@***.***> wrote:

> *yukosgiti* created an issue (vercel/next.js#91642)
> <https://github.com/vercel/next.js/issues/91642>
> Link to the code that reproduces this issue
>
> https://github.com/yukosgiti/next-js-16-2-0-bug
> To Reproduce
>
> The reproduction has 2 commits.
> 1st commit - pdfjs working normally in a server action.
> 2nd commit - ran npx next upgrade
>
>    - checkout the 1st commi
>    - npm install
>    - npm run dev
>    - click on button at /
>    - pdf created.
>    - then run: npx next upgrade
>    - the dev server errors
>
> Current vs. Expected behavior
>
> the website should work as it did before the upgrade
>
> The build and the dev server gives the following error:
>
> GET / 500 in 32ms (next.js: 19ms, application-code: 13ms)[browser] Uncaught Error: ./node_modules/fflate/lib/node.cjs:22:13Module not found: Can't resolve <dynamic>
>   20 | node_worker_1["default"] = Worker ? function (c, _, msg, transfer, cb) {
>   21 |     var done = false;> 22 |     var w = new Worker(c + workerAdd, { eval: true })
>      |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   23 |         .on('error', function (e) { return cb(e, null); })
>   24 |         .on('message', function (m) { return cb(null, m); })
>   25 |         .on('exit', function (c) {
>
>
>
> Import trace:
>   Server Component:
>     ./node_modules/fflate/lib/node.cjs
>     ./node_modules/jspdf/dist/jspdf.node.min.js
>     ./app/create-pdf-action.ts
>     ./app/page.tsx
>
> https://nextjs.org/docs/messages/module-not-found
>
>
>     at <unknown> (Error: ./node_modules/fflate/lib/node.cjs:22:13)
>     at <unknown> (https://nextjs.org/docs/messages/module-not-found)
>     at <unknown> (https://nextjs.org/docs/messages/module-not-found)[browser] ./node_modules/fflate/lib/node.cjs:22:13Module not found: Can't resolve <dynamic>
>   20 | node_worker_1["default"] = Worker ? function (c, _, msg, transfer, cb) {
>   21 |     var done = false;> 22 |     var w = new Worker(c + workerAdd, { eval: true })
>      |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   23 |         .on('error', function (e) { return cb(e, null); })
>   24 |         .on('message', function (m) { return cb(null, m); })
>   25 |         .on('exit', function (c) {
>
> Import trace:
>   Server Component:
>     ./node_modules/fflate/lib/node.cjs
>     ./node_modules/jspdf/dist/jspdf.node.min.js
>     ./app/create-pdf-action.ts
>     ./app/page.tsx
>
> https://nextjs.org/docs/messages/module-not-found
>     at console.overrideMethod [as error] (chrome-extension://fmkadmapgofadopljbjfkapdkoienihi/build/installHook.js:1:174822) (chrome-extension://fmkadmapgofadopljbjfkapdkoienihi/build/installHook.js:1:174822)[browser] ./node_modules/fflate/lib/node.cjs:22:13Module not found: Can't resolve <dynamic>
>   20 | node_worker_1["default"] = Worker ? function (c, _, msg, transfer, cb) {
>   21 |     var done = false;> 22 |     var w = new Worker(c + workerAdd, { eval: true })
>      |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   23 |         .on('error', function (e) { return cb(e, null); })
>   24 |         .on('message', function (m) { return cb(null, m); })
>   25 |         .on('exit', function (c) {
>
> Import trace:
>   Server Component:
>     ./node_modules/fflate/lib/node.cjs
>     ./node_modules/jspdf/dist/jspdf.node.min.js
>     ./app/create-pdf-action.ts
>     ./app/page.tsx
> https://nextjs.org/docs/messages/module-not-found
>     at console.overrideMethod [as error] (chrome-extension://fmkadmapgofadopljbjfkapdkoienihi/build/installHook.js:1:174822) (chrome-extension://fmkadmapgofadopljbjfkapdkoienihi/build/installHook.js:1:174822)
>
> Provide environment information
>
> Operating System:
>   Platform: darwin
>   Arch: arm64
>   Version: Darwin Kernel Version 25.3.0: Wed Jan 28 20:56:42 PST 2026; root:xnu-12377.91.3~2/RELEASE_ARM64_T8142
>   Available memory (MB): 16384
>   Available CPU cores: 10
> Binaries:
>   Node: 22.21.1
>   npm: 11.6.4
>   Yarn: 1.22.22
>   pnpm: 10.26.0
> Relevant Packages:
>   next: 16.2.0 // Latest available version is detected (16.2.0).
>   eslint-config-next: N/A
>   react: 19.2.4
>   react-dom: 19.2.4
>   typescript: 5.9.3
> Next.js Config:
>   output: N/A
>
> Which area(s) are affected? (Select all that apply)
>
> React, Server Actions
> Which stage(s) are affected? (Select all that apply)
>
> next dev (local), next build (local)
> Additional context
>
> It might be an issue with react as well.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/vercel/next.js/issues/91642>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/B67D5GO6X7OVTWKALDXPQT34RO5FJAVCNFSM6AAAAACWXWJPWOVHI2DSMVQWIX3LMV43ASLTON2WKOZUGEYDANBWGE3DGMI>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

same here:
```
> Build error occurred
Error: Turbopack build failed with 1 errors:
./node_modules/fflate/lib/node.cjs:22:13
Module not found: Can't resolve <dynamic>
  20 | node_worker_1["default"] = Worker ? function (c, _, msg, transfer, cb) {
  21 |     var done = false;
> 22 |     var w = new Worker(c + workerAdd, { eval: true })
     |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  23 |         .on('error', function (e) { return cb(e, null); })
  24 |         .on('message', function (m) { return cb(null, m); })
  25 |         .on('exit', function (c) {



Import trace:
  SSR:
    ./node_modules/fflate/lib/node.cjs
    ./node_modules/jspdf/dist/jspdf.node.min.js
```

--- a/test/e2e/app-dir/node-worker-threads/app/api/jspdf-test/route.ts
+++ b/test/e2e/app-dir/node-worker-threads/app/api/jspdf-test/route.ts
@@ -1,0 +1,20 @@
+import { NextResponse } from 'next/server'
+
+export async function GET() {
+  try {
+    const { jsPDF } = await import('jspdf')
+    const doc = new jsPDF()
+    doc.text('Hello world!', 10, 10)
+    const pdfData = doc.output('arraybuffer')
+
+    return NextResponse.json({
+      success: true,
+      size: pdfData.byteLength,
+    })
+  } catch (error) {
+    return NextResponse.json(
+      { success: false, error: String(error) },
+      { status: 500 }
+    )
+  }
+}

--- a/test/e2e/app-dir/node-worker-threads/node-worker-threads.test.ts
+++ b/test/e2e/app-dir/node-worker-threads/node-worker-threads.test.ts
@@ -6,6 +6,7 @@ describe('node-worker-threads', () => {
     skipDeployment: true,
     dependencies: {
       pino: '9.6.0',
+      jspdf: '4.2.1',
     },
   })
 
@@ -56,6 +57,17 @@ describe('node-worker-threads', () => {
     // The __turbopack_globals__ key should NOT be visible to user code
     expect(data.hasTurbopackKeys).toBe(false)
     expect(data.turbopackKeys).toEqual([])
+  })
+
+  it('should handle jsPDF which uses Worker with eval: true (issue #91642)', async () => {
+    // jsPDF internally creates Worker threads with { eval: true }, passing
+    // inline JS code instead of a file path. Turbopack should not try to
+    // resolve the first argument as a module reference in this case.
+    const res = await next.fetch('/api/jspdf-test')
+    const data = await res.json()
+    expect(res.status).toBe(200)
+    expect(data.success).toBe(true)
+    expect(data.size).toBeGreaterThan(0)
   })
 
   it('should handle PNG file import in worker', async () => {

--- a/turbopack/crates/turbopack-core/src/resolve/parse.rs
+++ b/turbopack/crates/turbopack-core/src/resolve/parse.rs
@@ -293,7 +293,12 @@ impl Request {
                 Request::Unknown { path } => {
                     path.push(item);
                 }
-                Request::DataUri { .. } | Request::Uri { .. } | Request::Dynamic => {
+                Request::DataUri { .. } | Request::Uri { .. } => {
+                    return Request::Dynamic;
+                }
+                Request::Dynamic => {
+                    // A dynamic prefix is essentially impossible to resolve so we don't try.  We
+                    // would have to scan the entire repo for suffix matches.
                     return Request::Dynamic;
                 }
                 Request::Alternatives { .. } => unreachable!(),

--- a/turbopack/crates/turbopack-ecmascript/src/references/mod.rs
+++ b/turbopack/crates/turbopack-ecmascript/src/references/mod.rs
@@ -2167,6 +2167,58 @@ where
             WellKnownFunctionKind::NodeWorkerConstructor => {
                 let args = linked_args().await?;
                 if !args.is_empty() {
+                    // When `{ eval: true }` is passed as the second argument,
+                    // the first argument is inline JS code, not a file path.
+                    // Skip creating a worker reference in that case.
+                    let mut dynamic_warning: Option<&str> = None;
+                    if let Some(opts) = args.get(1) {
+                        match opts {
+                            JsValue::Object { parts, .. } => {
+                                let eval_value = parts.iter().find_map(|part| match part {
+                                    ObjectPart::KeyValue(
+                                        JsValue::Constant(JsConstantValue::Str(key)),
+                                        value,
+                                    ) if key.as_str() == "eval" => Some(value),
+                                    _ => None,
+                                });
+                                if let Some(eval_value) = eval_value {
+                                    match eval_value {
+                                        // eval: true — first arg is code, not a
+                                        // path
+                                        JsValue::Constant(JsConstantValue::True) => {
+                                            return Ok(());
+                                        }
+                                        // eval: false — first arg is a path,
+                                        // continue normally
+                                        JsValue::Constant(JsConstantValue::False) => {}
+                                        // eval is set but not a literal boolean
+                                        _ => {
+                                            dynamic_warning = Some("has a dynamic `eval` option");
+                                        }
+                                    }
+                                }
+                            }
+                            // Options argument is not a static object literal —
+                            // we can't inspect it for `eval: true`
+                            _ => {
+                                dynamic_warning = Some("has a dynamic options argument");
+                            }
+                        }
+                    }
+                    if let Some(warning) = dynamic_warning {
+                        let (args, hints) = explain_args(args);
+                        handler.span_warn_with_code(
+                            span,
+                            &format!("new Worker({args}) {warning}{hints}"),
+                            DiagnosticId::Lint(
+                                errors::failed_to_analyze::ecmascript::NEW_WORKER.to_string(),
+                            ),
+                        );
+                        if ignore_dynamic_requests {
+                            return Ok(());
+                        }
+                    }
+
                     let pat = js_value_to_pattern(&args[0]);
                     if !pat.has_constant_parts() {
                         let (args, hints) = explain_args(args);
@@ -2187,12 +2239,13 @@ where
                     } else {
                         ResolveErrorMode::Error
                     };
+                    let pattern = Pattern::new(pat).to_resolved().await?;
                     analysis.add_reference_code_gen(
                         WorkerAssetReference::new_node_worker_thread(
                             origin,
                             // WorkerThreads resolve filepaths relative to the process root
                             get_traced_project_dir().await?,
-                            Pattern::new(pat).to_resolved().await?,
+                            pattern,
                             collect_affecting_sources,
                             get_issue_source(),
                             error_mode,
@@ -2208,7 +2261,7 @@ where
                     span,
                     &format!("new Worker({args}) is not statically analyze-able{hints}",),
                     DiagnosticId::Error(
-                        errors::failed_to_analyze::ecmascript::FS_METHOD.to_string(),
+                        errors::failed_to_analyze::ecmascript::NEW_WORKER.to_string(),
                     ),
                 );
                 // Ignore (e.g. dynamic parameter or string literal)

--- a/turbopack/crates/turbopack-ecmascript/src/references/mod.rs
+++ b/turbopack/crates/turbopack-ecmascript/src/references/mod.rs
@@ -2239,13 +2239,12 @@ where
                     } else {
                         ResolveErrorMode::Error
                     };
-                    let pattern = Pattern::new(pat).to_resolved().await?;
                     analysis.add_reference_code_gen(
                         WorkerAssetReference::new_node_worker_thread(
                             origin,
                             // WorkerThreads resolve filepaths relative to the process root
                             get_traced_project_dir().await?,
-                            pattern,
+                            Pattern::new(pat).to_resolved().await?,
                             collect_affecting_sources,
                             get_issue_source(),
                             error_mode,

